### DISCUSSION TITLE
Having trouble with simple configuration

### DISCUSSION THREAD

**Discussion by Dunge** [dotnet/yarp#1373] [Q&A] [answered]
Hello there!

This project is promising, I like the idea of having my main AspNetCore application running on Kestrel allowing to redirect just certain traffic to a different web server without a third party application and/or machine in the mix.

Unfortunately, I have to say your configuration documentation is severely lacking.  It barely list all possible config entries and redirect to the api documentation that shows the types and a short sentence, but it doesn't have any concrete examples.

I'm trying to redirect all traffic matching a certain Path and a certain subdomain (Host) to another IP. Here my config:

```json
  "ReverseProxy": {
    "Routes": {
      "route1" : {
        "ClusterId": "cluster1",
        "Match": {
          "Path": "/test/{*any}"
        }
      },
      "route2" : {
        "ClusterId": "cluster1",
        "Match": {
          "Hosts": [ "test.*" ],
        }
      }
    },
    "Clusters": {
      "cluster1": {
        "Destinations": {
          "destination1": {
            "Address": "https://localhost:5001/test"
          }
        },
        "HttpClient": {
          "DangerousAcceptAnyServerCertificate": "true"
        },
      }
    }
  },
```

Problem # 1: The path match works, but end up redirecting to /test/test:`

     2021-11-10 23:01:51.7130 Yarp.ReverseProxy.Forwarder.HttpForwarder Proxying to https://localhost:5001/test/test/ HTTP/2 RequestVersionOrLower no-streaming 

A bit confusing since it's the same name here, but the point is that the path to match might not be the same than the path to redirect to. I would have expected only the part inside the {*any} statement to be appended to the destination URL, not the whole path including the static matching word "test".

Problem # 2: Hosts redirection never get caught at all. I connect to test.mydomain.com or test.localhost and it doesn't seems to be handled and just show my normal website with no redirect entry in the logfile. I even tried with full domain "Hosts": [ "test.mydomain.com" ], nothing. Eventually I wish to have a wildcard so that it works on many domain without having to re-write the configuration.

**Comment by Tratcher** [comment, marked as answer]
1) The path isn't modified by the route match but you can use transforms for that. See https://microsoft.github.io/reverse-proxy/articles/transforms.html#pathpattern

2) "test.*" isn't a supported host routing wildcard pattern, only full domains or wildcard prefixes ("*.mydomain.com") are supported. "test.mydomain.com" should have worked. Can you share a fiddler trace of that request?

**Reply by Dunge** [reply]
1. Thanks, I managed to get it working with this:

```json
        "Transforms": [
          { "PathRemovePrefix": "/test/" }
        ]
```

2. So you are saying there's no way to configure a match for a subdomain name? Let's say I deploy the same application on two separate domains and want to match the same subdomain on both, I need a different configuration? Could I open a ticket for a future feature? Seems like something that is common and should be supported.

In any case, I wanted to keep my org name out of this discussion, but not too important. Here is the fiddler trace with the domain. Not sure what info you want to get out of this trace, it only shows the GET request and the result (which is the website default page, not the forwarded one). The forward itself should be transparent from Fiddler point of view. I tried both with  "Hosts": [ "test.localhost" ], and "Hosts": [ "test.dev-svr3.domain.com" ], nothing.

**Reply by Tratcher** [reply]
The host matching is implemented in aspnetcore routing so you can ask for the "test.*" pattern support at: https://github.com/dotnet/aspnetcore/issues

> Edit: Adding any path (/whatever) after the domain in the URL does redirect correctly, just not the plain domain URL itself (ending in .com).

Path matches are higher priority than host matches, it sounds like you have a route that matches "/" or "/{**remainder}" and that's taking precedence. Enabling aspnetcore debug level logging would show which route was matched for a request.

**Reply by Dunge** [reply]
@Tratcher My ReverseProxy config is completely pasted above, there's no additional route match. But after your comment, I wondered if the rest of the AspNetCore application route could influence, and yes it does.

Removing my MVC route fix the problem:

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapReverseProxy();
                //endpoints.MapControllerRoute("default", "{controller}/{action=Index}/{enterpriseId:int?}");
            });

This route is registered AFTER the reverse proxy registration, and have nothing to do with host/subdomain name. It's just for my normal website controller/action matching when there's no special redirection to be done.

I assume this is a bug? How can I have both working?

Still on .NET5 if that have any importance.

**Reply by Tratcher** [reply]
YARP integrates into the same aspnetcore routing table so requests could be routed to controllers or YARP. Registration order doesn't matter, the most specific path match will win, host matching will only be used as a tie-breaker. You can manually adjust the order of route precedence by setting the route Order field (lower/negative comes first).

You'll have to work out your route match rules so it's not ambiguous when a request should be handled by a controller and when it should be handled by YARP.

**Reply by Dunge** [reply]
How can I set YARP route to be prioritized over Asp controller routes? How can I set the order number in the MapControllerRoute method?

[Documentation ](https://docs.microsoft.com/en-us/aspnet/core/mvc/controllers/routing?view=aspnetcore-6.0)state:

> MapControllerRoute and MapAreaRoute :
> Automatically assign an order value to their endpoints based on the order they are invoked.

**Reply by Tratcher** [reply]
You can set the order in YARP like this:
https://github.com/microsoft/reverse-proxy/blob/b2932df6c37044f5a72a5ef9228cf43e6b64b858/samples/ReverseProxy.Config.Sample/appsettings.json#L32
I don't know which specific value you'll need to override controllers, but a large negative number should do.
https://github.com/microsoft/reverse-proxy/blob/b2932df6c37044f5a72a5ef9228cf43e6b64b858/src/ReverseProxy/Routing/HeaderMatcherPolicy.cs#L19

**Reply by Dunge** [reply]
Oh wow, high _negative_ number priority! Setting my YARP routes to -10001 and -10000 actually seems to have worked.

Thanks for all the help.

But honestly, this is highly convoluted. You really should check for a better method of doing this. It's really weird that we have to dig into source code to see default _negative_ priorities and chose a number below just to have YARP co-exist with ASP controllers..
